csrc/kernel_cache.cpp

```diff
@@ -1585,37 +1586,102 @@ void FusionKernelRuntime::updateHeuristicsLaunchParams(
 std::optional<FusionKernelRuntime::HeuristicsPtr> FusionKernelRuntime::
     getMaybeHeuristicsFor(
         const KernelArgumentHolder& args,
-        std::optional<PrimDataType> forced_index_type) {
+        std::optional<PrimDataType> forced_index_type,
+        bool initial_heuristics) {
   FUSER_PERF_SCOPE("FusionKernelRuntime::getMaybeHeuristicsFor");
-  auto complete_fusion = segmented_fusion_->completeFusion();
-  precomputed_values_->bindInputs(args);
-  precomputed_values_->evaluate();
-  SchedulerRuntimeInfo runtime_info(
-      complete_fusion,
-      args,
-      precomputed_values_.get(),
-      all_tvs_,
-      forced_index_type);
-
-  std::optional<FusionKernelRuntime::HeuristicsPtr> ret;
-  ret = std::make_unique<FusionHeuristics>();
-  size_t total_groups = segmented_fusion_->groups().size();
-  for (const auto group_index : c10::irange(total_groups)) {
-    auto group = segmented_fusion_->groups()[group_index];
-
-    auto maybe_scheduler_entry = group->getMaybeSchedulerEntry(runtime_info);
-    if (!maybe_scheduler_entry.has_value()) {
-      return std::nullopt;
+
+  // The runtime group run order is different from the segmented_fusion group
+  // order. Instead of using FusionHeuristics::emplaceBack, we initialize
+  // FusionHeuristics with the desired number of groups.
+  std::optional<FusionKernelRuntime::HeuristicsPtr> opt_heuristics;
+  const int64_t num_groups = (int64_t)runtime_workspace_.group_run_order.size();
+  opt_heuristics = std::make_unique<FusionHeuristics>(num_groups);
+
+  // Store metadata copy of arguments for ArgumentManager
+  KernelArgumentHolder args_metadata;
+  std::transform(
+      args.cbegin(),
+      args.cend(),
+      args_metadata.getBackInserter(),
+      convertMetadataArg);
+  args_metadata.setDeviceIndex(args.getDeviceIndex());
+
+  // ArgumentManager manipulates the KernelArgumentHolder argument.
+  // We make another metadata copy for the PrecomputedValues.
+  KernelArgumentHolder complete_fusion_metadata_args(args_metadata);
+
+  ArgumentManager args_manager(
+      args_metadata, runtime_workspace_, segmented_fusion_->inputs());
+
+  // Follow group run order
+  for (int64_t group_id = 0; group_id < num_groups; ++group_id) {
+    auto group_to_run = runtime_workspace_.group_run_order.at(group_id);
+
+    // Create fusion for this segmented group
+    Fusion* fusion_to_run = group_to_run->getFusion();
```

Comment:
Cleanup suggestions: given this cloning is for `makeInitialSchedulerEntry` and `getMaybeSchedulerEntry`, can we move cloning to those methods? This way, we don't need to add `fusion_to_run` to the interface of SegmentedFusion.

Reply:
`fusion_to_run` is required for `PrecomputedValues`, `SchedulerRuntimeInfo`, and `FusionExecutor::inferOutputSizes`, which are outside of `makeInitialSchedulerEntry`.